odl/trafos/backends/pywt_bindings.py

```diff
@@ -52,760 +52,101 @@ def pywt_wavelet(wavelet):
-        is the number of dimensions of the input array.
-    wavelet :  string or `pywt.Wavelet`
-        Specification of the wavelet to be used in the transform.
-        Use `pywt.wavelist` to get a list of available wavelets.
-    mode : string, optional
-        PyWavelets style signal extension mode. See `signal extension modes`_
-        for available options.
-    recon_shape : sequence of ints, optional
-        Shape of the array to be reconstructed. Without this parameter,
-        the reconstructed array always has even shape due to upsampling
-        by a factor of 2. To get reconstructions with odd shapes, this
-        parameter is required.
+    coeff_shapes : array-like
+        A list of multilevel wavelet coefficient shapes as returned by
+        `pywt.wavedecn_shapes`.
+    axes : sequence of ints, optional
+        Axes over which the DWT that created ``coeffs`` was performed. The
+        default value of None corresponds to all axes.
 
     Returns
     -------
-    recon : `numpy.ndarray`
-        Wavelet reconstruction from the given coefficients.
-
-    See Also
-    --------
-    pywt_single_level_recon : Single-level version of this function.
-    pywt_multi_level_decomp : Multi-level decomposition, i.e. the inverse
-        of this function.
+    coeff_slices : list
+        List of slices corresponding to each coefficient. As a 2D example,
+        ``coeff_arr[coeff_slices[1]['dd']]`` would extract the first level
+        detail coefficients from ``coeff_arr``.
 
     Examples
     --------
-    Reconstruct the original array from the decomposition in the example
-    in `pywt_multi_level_decomp`:
-
-    >>> orig_arr = [[1, 1, 0, 0],
-    ...             [0, 0, 0, 1],
-    ...             [1, 1, 1, 1],
-    ...             [0, 1, 1, 0]]
-    >>> approx2 = [[2.25]]
-    >>> details2 = ([[0.25]], [[-0.75]], [[0.25]])
-    >>> details1 = ([[0, -0.5],
-    ...              [-0.5, 0.5]],
-    ...             [[1, -0.5],
-    ...              [0.5, 0.5]],
-    ...             [[0, 0.5],
-    ...              [0.5, -0.5]])
-    >>> coeffs = [approx2, details2, details1]
-    >>> recon = pywt_multi_level_recon(coeffs, wavelet='haar', mode='zero')
-    >>> np.allclose(recon, orig_arr)
-    True
-
-    References
-    ----------
-    .. _signal extension modes:
-       https://pywavelets.readthedocs.io/en/latest/ref/signal-extension-\
-modes.html
+    >>> import pywt
+    >>> data_shape = (64, 64)
+    >>> coeff_shapes = pywt.wavedecn_shapes(data_shape, wavelet='db2', level=3,
+    ...                                     mode='periodization')
+    >>> coeff_slices = precompute_raveled_slices(coeff_shapes)
+    >>> print(coeff_slices[0])  # approximation coefficients
+    slice(None, 64, None)
+    >>> print(coeff_slices[-1])  # first level detail coefficients
+    {'ad': slice(1024, 2048, None), 'da': slice(2048, 3072, None), 'dd': slice(3072, 4096, None)}
```

Comment:
Maybe
```python
>>> d1_slices = coeff_slices[-1]  # first level detail coefficients
>>> d1_slices['ad'], d1_slices['da'], d1_slices['dd']
(slice(1024, 2048, None), slice(2048, 3072, None), slice(3072, 4096, None))
```

Reply:
But the version after your edit is also fine. This function is somewhat low-level anyway, so the target audience here is more the developer than the user.